homeassistant/components/discovergy/sensor.py

```diff
@@ -0,0 +1,274 @@
+        suggested_display_precision=1,
+        device_class=SensorDeviceClass.VOLTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        entity_registry_enabled_default=False,
+    ),
+    # energy sensors
+    DiscovergySensorEntityDescription(
+        key="energy",
+        translation_key="total_consumption",
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        suggested_display_precision=4,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+        scale=10000000000,
+    ),
+    DiscovergySensorEntityDescription(
+        key="energyOut",
+        translation_key="total_production",
+        native_unit_of_measurement=UnitOfEnergy.KILO_WATT_HOUR,
+        suggested_display_precision=4,
+        device_class=SensorDeviceClass.ENERGY,
+        state_class=SensorStateClass.TOTAL_INCREASING,
+        scale=10000000000,
+    ),
+)
+
+
+def get_coordinator_for_meter(
+    hass: HomeAssistant,
+    meter: Meter,
+    discovergy_instance: Discovergy,
+    update_interval: timedelta,
+) -> DataUpdateCoordinator:
+    """Create a new DataUpdateCoordinator for given meter."""
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            return await discovergy_instance.get_last_reading(meter.get_meter_id())
+        except AccessTokenExpired as err:
+            raise ConfigEntryAuthFailed(
+                "Got token expired while communicating with API"
+            ) from err
+        except HTTPError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+        except Exception as err:  # pylint: disable=broad-except
+            raise UpdateFailed(
+                f"Unexpected error while communicating with API: {err}"
+            ) from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=update_interval,
+    )
+    return coordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Discovergy sensors."""
+    data: DiscovergyData = hass.data[DOMAIN][entry.entry_id]
+    discovergy_instance: Discovergy = data.api_client
+    meters: list[Meter] = data.meters  # always returns a list
+
+    min_time_between_updates = timedelta(
+        seconds=entry.options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
+    )
+
+    entities = []
+    for meter in meters:
+        # Get coordinator for meter, set config entry and fetch initial data
+        # so we have data when entities are added
+        coordinator = get_coordinator_for_meter(
+            hass, meter, discovergy_instance, min_time_between_updates
+        )
+        coordinator.config_entry = entry
+        await coordinator.async_config_entry_first_refresh()
```

Comment:
Setting up the coordinators should move to the config entry setup in the init module. Otherwise this method call won't work properly with exceptions raised etc.